### PR TITLE
[Translation] Added support for translation files with other filename patterns

### PR DESCRIPTION
**Pull request by javiereguiluz**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #27644
| License       | MIT
| Doc PR        | -

This implements the #27644 feature request in case we accept it.

My vote is 👍 because the changes required are tiny and the resulting code is even more robust thanks to the new `preg_match()` call.

### CODE REVIEW

File: src/Symfony/Component/Translation/Command/XliffLintCommand.php [hunk 1]
```diff
@@ -115,14 +115,14 @@ private function validate($content, $file = null)
         $document->loadXML($content);
 
         if (null !== $targetLanguage = $this->getTargetLanguageFromFile($document)) {
-            $expectedFileExtension = sprintf('%s.xlf', str_replace('-', '_', $targetLanguage));
-            $realFileExtension = explode('.', basename($file), 2)[1] ?? '';
+            $normalizedLocale = str_replace('-', '_', $targetLanguage);
+            $expectedFilenamePattern = sprintf('/^(.*\.%s\.xlf|%s\..*\.xlf)/', $normalizedLocale, $normalizedLocale);
```
Comment (stof) [inline src/Symfony/Component/Translation/Command/XliffLintCommand.php]: missing `preg_quote` on the 2 variables you inject in the regex

**Review by stof — CHANGES_REQUESTED** [review]

**Comment by stof** [comment]
In a FrameworkBundle-based project, the file name format must always be `<domain>.<locale>.<format>`. Anything else is unsupported (because the bundle needs to be able to extract the domain and the locale from the file name). So having the convention configurable in the command (using a constructor argument) would be better to me than supporting having the locale anywhere in the file name (which would not report all issues for Symfony projects)

**Comment by javiereguiluz** [comment]
@stof I've made some changes. Please tell me if that's what you were proposing. Thanks.

File: src/Symfony/Component/Translation/Command/XliffLintCommand.php [hunk 2]
```diff
@@ -115,14 +117,16 @@ private function validate($content, $file = null)
         $document->loadXML($content);
 
         if (null !== $targetLanguage = $this->getTargetLanguageFromFile($document)) {
-            $expectedFileExtension = sprintf('%s.xlf', str_replace('-', '_', $targetLanguage));
-            $realFileExtension = explode('.', basename($file), 2)[1] ?? '';
+            $normalizedLocale = preg_quote(str_replace('-', '_', $targetLanguage));
+            // when this command is used inside Symfony, require filenames to be '____.locale.xlf'
+            // when used outside Symfony, allow '____.locale.xlf' and 'locale.____.xlf'
```
Comment (stof) [inline src/Symfony/Component/Translation/Command/XliffLintCommand.php]: comments should be talking about strict vs not strict, not about `inside Symfony`. Other projects may put the command in strict mode too.

File: src/Symfony/Component/Translation/Command/XliffLintCommand.php [hunk 3]
```diff
@@ -33,13 +33,15 @@ class XliffLintCommand extends Command
     private $displayCorrectFiles;
     private $directoryIteratorProvider;
     private $isReadableProvider;
+    private $requireStrictFileNames;
 
-    public function __construct(string $name = null, callable $directoryIteratorProvider = null, callable $isReadableProvider = null)
+    public function __construct(string $name = null, callable $directoryIteratorProvider = null, callable $isReadableProvider = null, bool $requireStrictFileNames = false)
```
Comment (stof) [inline src/Symfony/Component/Translation/Command/XliffLintCommand.php]: Shouldn't we keep using `true` as default, to keep the same behavior than today by default ?

File: src/Symfony/Component/Translation/Command/XliffLintCommand.php [hunk 4]
```diff
@@ -115,14 +117,16 @@ private function validate($content, $file = null)
         $document->loadXML($content);
 
         if (null !== $targetLanguage = $this->getTargetLanguageFromFile($document)) {
-            $expectedFileExtension = sprintf('%s.xlf', str_replace('-', '_', $targetLanguage));
-            $realFileExtension = explode('.', basename($file), 2)[1] ?? '';
+            $normalizedLocale = preg_quote(str_replace('-', '_', $targetLanguage));
```
Comment (stof) [inline src/Symfony/Component/Translation/Command/XliffLintCommand.php]: you need to pass the regex delimiter as second argument to `preg_quote` as it needs to be escaped too.

**Review by fabpot — COMMENTED** [review]
Can we add some more tests , especially when `requireStrictFileNames` is false?

**Review by fabpot — APPROVED** [review]

**Pull request merged**